### PR TITLE
Implement basic lobby gameplay

### PR DESCRIPTION
**Pull request by p2r3**
This pull request makes lobbies somewhat functional, albeit basic and limited. It is possible to create a lobby, set the map, start the game, and "submit" times.

I just want to get _something_ upstream, so that people can play around with it and give me additional motivation to continue working on this. This shouldn't affect the rest of Epochtal, and only some of the code is a placeholder implementation - mainly UI stuff and run submissions, the latter of which is marked with `// TODO` in the section of code responsible for that.

### CODE REVIEW

**Review by PancakeTAS — CHANGES_REQUESTED** [review]
This is what I attempted to do locally:
1. Create a lobby and paste gelocity into the map selection
2. Launch the game, type connect
3. Close the menu, type connect *again*, enter the code on the website like requested.
4. The popup for authenticating connection was now visible, I clicked "Okay" and clicked "I'm ready" on the lobby, which did absolutely nothing.
<details>
  <summary>Console log</summary>
  <code>
[Sun, 15 Sep 2024 14:56:52 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:57:04 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:57:16 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:57:28 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:57:40 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:57:52 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:58:04 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
[Sun, 15 Sep 2024 14:58:16 GMT UTC+0]
Error: ERR_TIMEOUT
    at new UtilError (/home/pancake/dev/vscode/epochtal/util/error.js:24:5)
    at <anonymous> (/home/pancake/dev/vscode/epochtal/util/lobbies.js:324:24)
- Command: "lobbies ready cute puppy girls :3 true 76561198870771271"
- Context: "epochtal"
  </code>
</details>
(idk what happened to the formatting)

**Comment by p2r3** [comment]
@PancakeTAS Just to make sure, confirm that this _only_ happens when the game client is authenticated twice? As in, the "expected" setup functions correctly?

File: main.js [hunk 1]
```diff
@@ -167,6 +167,11 @@ const fetchHandler = async function (req) {
   // Handle WebSocket connections
   if (urlPath[0] === "ws") {
 
+    // Load authentication header from parameters if present
+    // This is a workaround for internal requests, as the JavaScript WebSockets API doesn't support setting headers
+    const authSecret = url.searchParams.get("Authentication");
+    if (authSecret) req.headers.set("Authentication", authSecret);
+
```
Comment (soni801) [inline main.js:174]: Putting secrets in URL parameters is generally strongly discouraged, as the URL may not be encrypted even if using TLS.

Quick read about the topic:
https://www.fullcontact.com/blog/2016/04/29/never-put-secrets-urls-query-parameters/
<sub>(I havent read this exact article in its entirety, but skimming over it, it seems to get the idea)</sub>

File: pages/live/lobby.js [hunk 2]
```diff
@@ -207,26 +275,26 @@ async function lobbyInit () {
...
-          default: return showPopup("Unknown error", "The server returned an unexpected response: " + data, POPUP_ERROR);
+        case "ERR_LOGIN": return showPopup("Not logged in", "Please log in via Steam before editing lobby details.", POPUP_ERROR);
+        case "ERR_STEAMID": return showPopup("Unrecognized user", "Your SteamID is not present in the users database. WTF?", POPUP_ERROR);
+        case "ERR_NAME": return showPopup("Lobby not found", "An open lobby with this name does not exist.", POPUP_ERROR);
+        case "ERR_NEWNAME": return showPopup("Invalid lobby name", "Please keep the lobby name to 50 characters or less.", POPUP_ERROR);
+        case "ERR_EXISTS": return showPopup("Lobby name taken", "A lobby with this name already exists.", POPUP_ERROR);
+        case "ERR_PERMS": return showPopup("Permission denied", "You do not have permission to perform this action.", POPUP_ERROR);
 
-        }
+        default: return showPopup("Unknown error", "The server returned an unexpected response: " + requestData, POPUP_ERROR);
```
Comment (soni801) [inline pages/live/lobby.js:297]: This particular (or _very_ similar) switch statement(s) gets repeated _a lot_. Consider splitting this into its own function.

File: routines/epochtal.js [hunk 3]
```diff
@@ -265,8 +265,6 @@ async function releaseMap (context) {
 
     fs.renameSync(vmf, `${context.file.vmfs}/${context.data.week.map.id}.vmf.xz`);
 
-    context.data.week.map.file = portal2.map[0];
-
```
Comment (soni801) [inline routines/epochtal.js:269]: This is probably just me not understanding the code base, but this seems like a somewhat random line to remove?

**Review by soni801 — CHANGES_REQUESTED** [review]
So, I did encounter a few issues testing this for the first time:

**First issue**:
I get the following error after inputting the connection code on the website when it gives the success message:
```
epochtal-1  | [Sun, 15 Sep 2024 21:40:10 GMT UTC+0]
epochtal-1  | TypeError: undefined is not an object (evaluating 'gameauth[steamid].authcode')
epochtal-1  |     at <anonymous> (/app/util/gameauth.js:45:20)
epochtal-1  |     at <anonymous> (/app/util/gameauth.js:17:34)
epochtal-1  |     at <anonymous> (/app/api/gameauth.js:82:20)
epochtal-1  |     at processTicksAndRejections (:12:39)
epochtal-1  | - Command: "gameauth verify 3122"
epochtal-1  | - Context: "epochtal"
```
I'm logged into the website, steam is running, and I have one run in the main category on this instance (i did not create a new instance when testing, but used my previous one which shouldnt be fucked) - Everything should be "good" user wise, for context.

**Second issue**:
(caused by not having the trailing `/` in the URL as mentioned in Discord - kinda "fixable" but also kinda critical imo - 99% of people navigating to a website don't put the trailing slash there...)
The webinterface (`http://localhost:8080/live`) will say "Loading available lobbies..." seemingly forever if there are no lobbies. Accompanied by the following errors in the DevTools console:
image
This continues on, also throwing `Uncaught ReferenceError: createLobbyPopup is not defined` when trying to create a lobby.

**Third issue**:
Getting `Uncaught ReferenceError: selectLobbyMap is not defined` when immediately pressing "choose map" after creating a lobby. Trying to refresh logs the following (i named the lobby "cringe"):
```
epochtal-1  | [Sun, 15 Sep 2024 22:01:37 GMT UTC+0]
epochtal-1  | Error: ERR_NAME
epochtal-1  |     at new UtilError (/app/util/error.js:24:5)
epochtal-1  |     at <anonymous> (/app/util/lobbies.js:42:46)
epochtal-1  |     at <anonymous> (/app/util/lobbies.js:23:34)
epochtal-1  |     at <anonymous> (/app/api/lobbies.js:86:31)
epochtal-1  |     at processTicksAndRejections (:12:39)
epochtal-1  | - Command: "lobbies get cringe"
epochtal-1  | - Context: "epochtal"
```
Going back to the lobby list, the lobby is no longer there. I try to create a new lobby, exact same thing happens. (the lobby stops existing when i refresh the lobby subpage)

Side note: why do you open the lobby in a new tab when it's created??????? Why would anyone ever want that????????????????

---

So it's kinda broken. Idk what steps i "should" do to test it, but realistically as long as it doesnt work "out of the box", i don't think it's ready for merging. And even if i did something wrong (i probably did), it shouldn't be this easy to fuck it up.

**Comment by p2r3** [comment]
@soni801 Something's very off here, especially with the third issue. A refresh should _not_ immediately close the lobby, but rather bring you back to the list page. This is old behavior, before this PR. Given that the select map button also points to a missing function... are you sure you're not just testing on a stale container? One that has the current upstream code? This would explain everything you've encountered.

**Comment by soni801** [comment]
Note to future self (and anyone else): use the `--build` flag with `docker compose up` when testing. :YEP:

File: pages/live/lobby.js (hunk 2 above)
Comment (p2r3) [reply pages/live/lobby.js:297]: Considered, but for a bigger frontend refactor. API queries like these are everywhere, and the handling is different in every file. Should probably be a part of `tools.js`. My point is, it's outside of the scope for this PR.

File: routines/epochtal.js (hunk 3 above)
Comment (p2r3) [reply routines/epochtal.js:269]: This is a part of commit 8e686881da66b97d0d63a9919172d2f718e232a4, which made the `file` property universal for all map data objects retrieved from the `workshopper` utility. The file path is required for lobbies to determine if the players have downloaded the map. Without this change, we'd need to duplicate a lot of code to achieve the same thing. Point is, this line simply became redundant.

I should probably take a step back at some point to see how much else was obsoleted with this change, but for now, I'll keep changes to the rest of the system minimal.

**Review by soni801 — CHANGES_REQUESTED** [review]
New review time...

I still get the following log error when verifying the game connection:
```
epochtal-1  | [Sun, 15 Sep 2024 22:32:28 GMT UTC+0]
epochtal-1  | TypeError: undefined is not an object (evaluating 'gameauth[steamid].authcode')
epochtal-1  |     at <anonymous> (/app/util/gameauth.js:45:20)
epochtal-1  |     at <anonymous> (/app/util/gameauth.js:17:34)
epochtal-1  |     at <anonymous> (/app/api/gameauth.js:91:20)
epochtal-1  |     at processTicksAndRejections (:12:39)
epochtal-1  | - Command: "gameauth verify 0624"
epochtal-1  | - Context: "epochtal"
```
However, the website still reports a success. And it seems to be working?

The issue with the missing trailing slash is also still there - I wouldn't ship this without at the very least redirecting the user to the page with a trailing slash. 99% of people don't write the trailing slash there, or even think of it as being a different path.

---

Now I seem to also be getting this error (when choosing map and immediately pressing "I'm ready"):
```
epochtal-1  | [Sun, 15 Sep 2024 22:36:41 GMT UTC+0]
epochtal-1  | Error: ERR_TIMEOUT
epochtal-1  |     at new UtilError (/app/util/error.js:24:5)
epochtal-1  |     at <anonymous> (/app/util/lobbies.js:324:24)
epochtal-1  | - Command: "lobbies ready cringe true 76561198845057322"
epochtal-1  | - Context: "epochtal"
epochtal-1  |
epochtal-1  | [Sun, 15 Sep 2024 22:36:54 GMT UTC+0]
epochtal-1  | Error: ERR_EXISTS
epochtal-1  |     at new UtilError (/app/util/error.js:24:5)
epochtal-1  |     at <anonymous> (/app/util/events.js:32:37)
epochtal-1  |     at <anonymous> (/app/util/events.js:19:34)
epochtal-1  |     at <anonymous> (/app/api/gameauth.js:62:15)
epochtal-1  |     at <anonymous> (/app/api/gameauth.js:16:34)
epochtal-1  |     at <anonymous> (/app/main.js:212:22)
epochtal-1  |     at fetchHandler (/app/main.js:160:38)
epochtal-1  | - Command: "events create game_76561198845057322 async function(steamid) {
epochtal-1  |           if (steamid === "00000000000000000")
epochtal-1  |             return !0;
epochtal-1  |           return epochtal.data.users[steamid].admin;
epochtal-1  |         } async function(message) {
epochtal-1  |           await events(["send", event, { type: "echo", value: message }]);
epochtal-1  |         } function() {
epochtal-1  |           const interval = setInterval(async function() {
epochtal-1  |             try {
epochtal-1  |               await events(["send", event, { type: "ping" }]);
epochtal-1  |             } catch {
epochtal-1  |               clearInterval(interval);
epochtal-1  |             }
epochtal-1  |           }, 30000);
epochtal-1  |         } async function() {
epochtal-1  |           await events(["delete", event]);
epochtal-1  |         }"
epochtal-1  | - Context: "epochtal"
```
(yes, that is two errors. I don't think i did anything special right after)

The website tells me that the request timed out and asks me to reconnect. That I do, this time with no log errors! (now I need to recreate the lobby because I spent too long on this so it deleted itself). Choosing map is fine, but I get the same ERR_TIMEOUT when pressing the ready button... idk again im probably doing something wrong but at least this time its the correct container lmfao

So no, I do not think this is ready to ship

File: pages/live/lobby.js (hunk 2 above)
Comment (soni801) [reply pages/live/lobby.js:297]: Understandable, but a temporary function would still be cleaner than repetition imo

**Comment by PancakeTAS** [comment]
connection works fine now, however after completing the map the time is not displayed in the browser without a reload of the page..

**Comment by PancakeTAS** [comment]
On an HTTPS host the lobby creation dialog prompts for username and password autofill on firefox :P

**Comment by PancakeTAS** [comment]
image
It doesn't work the first time. It authenticates, I get added to the lobby and the game immediately shows "WebSocket connection closed". After that, I connect again and this time it does work.

**Comment by PancakeTAS** [comment]
image
And once again the time is not updated and refreshing actually kicks me out of the lobby entirely

**Comment by p2r3** [comment]
1. Are you sure the token didn't just expire? They have a lifetime of 30 seconds. 
2. Refreshing is _supposed_ to kick you out of the lobby entirely, not sure why it didn't do that the last time you tested.
3. The time only updates once you exit the map, which it's supposed to do by default once the PTI end condition is met, though I'm now realizing that the active week's map is an exception here. That would make Gelocity not work on a testing build. Shouldn't be too hard to fix, though.

**Comment by p2r3** [comment]
Though, on second thought, maybe we should just outright prevent the active Epochtal map from being selected for the lobby? It'd be blatant route sharing if people did that, anyway.

File: defaults/portal2/main.js [hunk 4]
```diff
@@ -25,35 +28,46 @@ function wsMessageHandler (event) {
...
-    const protocol = SERVER_ADDRESS.startsWith("https:") ? "wss" : "ws";
-    const hostname = SERVER_ADDRESS.split("://")[1].split("/")[0];
-    const authcode = generateAuthCode();
+  const protocol = SERVER_ADDRESS.startsWith("https:") ? "wss" : "ws";
+  const hostname = SERVER_ADDRESS.split("://")[1].split("/")[0];
 
-    SendToConsole("hideconsole");
-    SendToConsole(`disconnect "Go to ${hostname} and enter the code ${authcode} when prompted."`);
+  if (ws) ws.close();
+  ws = new WebSocket(`${protocol}://${hostname}/api/events/connect`);
```
Comment (soni801) [inline defaults/portal2/main.js:59]: Wouldn't it be possible to just use `SERVER_ADDRESS`, but replace `http` with `ws`? lol

Sliiiiightly shorter implementation

**Review by soni801 — APPROVED** [review]
Shit worked, should be fine for initial release me thinks

File: defaults/portal2/main.js (hunk 4 above)
Comment (p2r3) [reply defaults/portal2/main.js:59]: I might be stupid

**Comment by PancakeTAS** [comment]
> 1. Are you sure the token didn't just expire? They have a lifetime of 30 seconds.
> 
> 2. Refreshing is _supposed_ to kick you out of the lobby entirely, not sure why it didn't do that the last time you tested.
> 
> 3. The time only updates once you exit the map, which it's supposed to do by default once the PTI end condition is met, though I'm now realizing that the active week's map is an exception here. That would make Gelocity not work on a testing build. Shouldn't be too hard to fix, though.

1. Seems to just have run out
2. Okay good, I guess
3. Seems good to me

I'm still annoyed at the way the gameauth works..

I created a lobby and didn't have my game open yet, so when it asked me to sync I simply couldn't. There's no mention of a 30 second token lifetime, so the average user would start their game now and the auth would not work. There is no message for the auth failing either, so the user thinks they are authenticated but they aren't.
I knew this, so I closed the window, started my game and wanted to link myself. Where do I click to do that? Nowhere, I have to pick a map and then click ready to get that popup again.
I picked a map from the workshop, clicked ready and then I authenticated successfully. Then I got an error for the map not being subscribed. I opened the page, subscribed and tried playing again.. same error. Turns out it didn't automatically download the map and required me to restart Portal 2. I restarted the game, the map was now downloaded, but I'm unable to relink again. I clicked "Ready" and the lobby just froze, presumably expecting my old client which was no longer running to join and play the game.
This isn't very user intuitive... at all.
Here's my suggestions:
1. Drop the gameauth token thing, it's stupid. The end user has Spplice running and the website open in their browser, likely even the steam browser. The end user should not have to figure out how to open the console (many users can't open the console in the main menu, because valve's stupid keybinds). There is *some* information that can be used to identify a user based on just the game and just the browser. It would be much cooler if the client had a main menu option for "Multiplayer -> Epochtal Live" or something like that, with which the client would repeatedly try to find the lobby they are in. There's no need to really worry about spoofing, because without the steam login token on the webpage the game is never able to start with a fake spoofed steamid or smth.
2. Automatically subscribe to the maps. I'm sure there is some steam weblink to automatically subscribe to something as I've actually seen some games like Civ6 use it. In the worst case, the server could send the map link via websocket to the clients and the js interface could download it.
3. There still seems to be a bug with restarting the game, which is probably the most urgent thing to fix right now. If you don't want to drop gameauth at least extend the token lifetime by.. a lot.

**Comment by p2r3** [comment]
1. We're not dropping tokens for authenticating with WebSockets. Assuming that a secure connection will never be needed is plain silly. We'd also have to make game authentication a complete exception to how WebSockets work elsewhere, which is doubly silly. The issues you presented are ones concerning UX, and should be solved _by improving UX_, not by dumbing down the implementation. We can add a "Token expired" message in Portal 2. We can add a button for fetching a new token on the site. We can present this to the user more intuitively. However, I refuse to loosely fingerprint the game client as a means of authentication.

2. Soni mentioned this on Discord. Once lobbies get more attention, we can outright download the bsp to the player's tempcontent if the map is missing. This seems to be the best solution, as subscribing to workshop maps sometimes requires a game restart. Automatically subscribing to the map for the user requires OAuth, and currently the user can just do so with two clicks within their browser, where they're already logged into Steam.

3. The "bug" with game restarts is a bug with Spplice 2, where the JS remains active after the game is closed. The request from the lobby does time out and tell you to reconnect. As for extending token lifetimes, any issues this causes are issues of UX - refer to #1.

**Comment by p2r3** [comment]
@PancakeTAS I've made some adjustments to make the token system more transparent and easier to work with. I've also implemented a workaround to Spplice leaving the socket open when the game closes. These changes have been tested. The workshop map download thing is on the roadmap, but not for the initial release.

Can you re-review? :pleading_face:

**Comment by p2r3** [comment]
I lied. The workshop map download thingy is now in the initial release.

File: defaults/portal2/main.js [hunk 5]
```diff
@@ -25,35 +25,102 @@ function wsMessageHandler (event) {
...
   }
 
 }
 
 // Authenticates and sets up a WebSocket connection
-function wsSetup () {
+function wsSetup (token) {
+
+  const protocol = SERVER_ADDRESS.startsWith("https:") ? "wss" : "ws";
+  const hostname = SERVER_ADDRESS.split("://")[1].split("/")[0];
```
Comment (PancakeTAS) [inline defaults/portal2/main.js:56]: This works.. but we do have the hostname in the configuration, so why not write that into address.txt?

**Review by PancakeTAS — APPROVED** [review]

**Pull request merged**